File: src/data_formats/parsing/ninja/nj.ts
import Logger from "js-logger";
import { BufferCursor } from "../../BufferCursor";
import { Vec3 } from "../../Vec3";
import { NinjaVertex } from "../ninja";

const logger = Logger.get("data_formats/parsing/ninja/nj");

// TODO:
// - textures
// - colors
// - bump maps
// - animation
// - deal with vertex information contained in triangle strips

export type NjModel = {
    type: "nj";
    /**
     * Sparse array of vertices.
     */
    vertices: NinjaVertex[];
    meshes: NjTriangleStrip[];
    // materials: [],
    bounding_sphere_center: Vec3;
    bounding_sphere_radius: number;
};

enum NjChunkType {
    Unknown,
    Null,
    Bits,
    CachePolygonList,
    DrawPolygonList,
    Tiny,
    Material,
    Vertex,
    Volume,
    Strip,
    End,
}

type NjChunk = {
    type: NjChunkType;
    type_id: number;
} & (
    | NjUnknownChunk
    | NjNullChunk
    | NjBitsChunk
    | NjCachePolygonListChunk
    | NjDrawPolygonListChunk
    | NjTinyChunk
    | NjMaterialChunk
    | NjVertexChunk
    | NjVolumeChunk
    | NjStripChunk
    | NjEndChunk);

type NjUnknownChunk = {
    type: NjChunkType.Unknown;
};

type NjNullChunk = {
    type: NjChunkType.Null;
};

type NjBitsChunk = {
    type: NjChunkType.Bits;
};

type NjCachePolygonListChunk = {
    type: NjChunkType.CachePolygonList;
    cache_index: number;
    offset: number;
};

type NjDrawPolygonListChunk = {
    type: NjChunkType.DrawPolygonList;
    cache_index: number;
};

type NjTinyChunk = {
    type: NjChunkType.Tiny;
};

type NjMaterialChunk = {
    type: NjChunkType.Material;
};

type NjVertexChunk = {
    type: NjChunkType.Vertex;
    vertices: NjVertex[];
};

type NjVolumeChunk = {
    type: NjChunkType.Volume;
};

type NjStripChunk = {
    type: NjChunkType.Strip;
    triangle_strips: NjTriangleStrip[];
};

type NjEndChunk = {
    type: NjChunkType.End;
};

type NjVertex = {
    index: number;
    position: Vec3;
    normal?: Vec3;
    bone_weight: number;
    bone_weight_status: number;
    calc_continue: boolean;
};

type NjTriangleStrip = {
    ignore_light: boolean;
    ignore_specular: boolean;
    ignore_ambient: boolean;
    use_alpha: boolean;
    double_side: boolean;
    flat_shading: boolean;
    environment_mapping: boolean;
    clockwise_winding: boolean;
    vertices: NjMeshVertex[];
};

type NjMeshVertex = {
    index: number;
    normal?: Vec3;
};

export function parse_nj_model(cursor: BufferCursor, cached_chunk_offsets: number[]): NjModel {
    const vlist_offset = cursor.u32(); // Vertex list
    const plist_offset = cursor.u32(); // Triangle strip index list
    const bounding_sphere_center = new Vec3(cursor.f32(), cursor.f32(), cursor.f32());
    const bounding_sphere_radius = cursor.f32();
    const vertices: NinjaVertex[] = [];
    const meshes: NjTriangleStrip[] = [];

    if (vlist_offset) {
        cursor.seek_start(vlist_offset);

        for (const chunk of parse_chunks(cursor, cached_chunk_offsets, true)) {
            if (chunk.type === NjChunkType.Vertex) {
                for (const vertex of chunk.vertices) {
                    vertices[vertex.index] = {
                        position: vertex.position,
                        normal: vertex.normal,
                        bone_weight: vertex.bone_weight,
                        bone_weight_status: vertex.bone_weight_status,
                        calc_continue: vertex.calc_continue,
                    };
                }
            }
        }
    }

    if (plist_offset) {
        cursor.seek_start(plist_offset);

        for (const chunk of parse_chunks(cursor, cached_chunk_offsets, false)) {
            if (chunk.type === NjChunkType.Strip) {
                meshes.push(...chunk.triangle_strips);
            }
        }
    }

    return {
        type: "nj",
        vertices,
        meshes,
        bounding_sphere_center,
        bounding_sphere_radius,
    };
}

// TODO: don't reparse when DrawPolygonList chunk is encountered.
function parse_chunks(
    cursor: BufferCursor,
    cached_chunk_offsets: number[],
    wide_end_chunks: boolean
): NjChunk[] {
    const chunks: NjChunk[] = [];
    let loop = true;

    while (loop) {
        const type_id = cursor.u8();
        const flags = cursor.u8();
        const chunk_start_position = cursor.position;
        let size = 0;

        if (type_id === 0) {
            chunks.push({
                type: NjChunkType.Null,
                type_id,
            });
        } else if (1 <= type_id && type_id <= 3) {
            chunks.push({
                type: NjChunkType.Bits,
                type_id,
            });
        } else if (type_id === 4) {
            const cache_index = flags;
            const offset = cursor.position;
            chunks.push({
                type: NjChunkType.CachePolygonList,
                type_id,
                cache_index,
                offset,
            });
            cached_chunk_offsets[cache_index] = offset;
            loop = false;
        } else if (type_id === 5) {
            const cache_index = flags;
            const cached_offset = cached_chunk_offsets[cache_index];

            if (cached_offset != null) {
                cursor.seek_start(cached_offset);
                chunks.push(...parse_chunks(cursor, cached_chunk_offsets, wide_end_chunks));
            }

            chunks.push({
                type: NjChunkType.DrawPolygonList,
                type_id,
                cache_index,
            });
        } else if (8 <= type_id && type_id <= 9) {
            size = 2;
            chunks.push({
                type: NjChunkType.Tiny,
                type_id,
            });
        } else if (17 <= type_id && type_id <= 31) {
            size = 2 + 2 * cursor.u16();
            chunks.push({
                type: NjChunkType.Material,
                type_id,
            });
        } else if (32 <= type_id && type_id <= 50) {
            size = 2 + 4 * cursor.u16();
            chunks.push({
                type: NjChunkType.Vertex,
                type_id,
                vertices: parse_vertex_chunk(cursor, type_id, flags),
            });
        } else if (56 <= type_id && type_id <= 58) {
            size = 2 + 2 * cursor.u16();
            chunks.push({
                type: NjChunkType.Volume,
                type_id,
            });
        } else if (64 <= type_id && type_id <= 75) {
            size = 2 + 2 * cursor.u16();
            chunks.push({
                type: NjChunkType.Strip,
                type_id,
                triangle_strips: parse_triangle_strip_chunk(cursor, type_id, flags),
            });
        } else if (type_id === 255) {
            size = wide_end_chunks ? 2 : 0;
            chunks.push({
                type: NjChunkType.End,
                type_id,
            });
            loop = false;
        } else {
            size = 2 + 2 * cursor.u16();
            chunks.push({
                type: NjChunkType.Unknown,
                type_id,
            });
            logger.warn(`Unknown chunk type ${type_id} at offset ${chunk_start_position}.`);
        }

        cursor.seek_start(chunk_start_position + size);
    }

    return chunks;
}

function parse_vertex_chunk(
    cursor: BufferCursor,
    chunk_type_id: number,
    flags: number
): NjVertex[] {
    if (chunk_type_id < 32 || chunk_type_id > 50) {
        logger.warn(`Unknown vertex chunk type ${chunk_type_id}.`);
        return [];
    }

    const bone_weight_status = flags & 0b11;
    const calc_continue = (flags & 0x80) !== 0;

    const index = cursor.u16();
    const vertex_count = cursor.u16();

    const vertices: NjVertex[] = [];

    for (let i = 0; i < vertex_count; ++i) {
        const vertex: NjVertex = {
            index: index + i,
            position: new Vec3(
                cursor.f32(), // x
                cursor.f32(), // y
                cursor.f32() // z
            ),
            bone_weight: 1,
            bone_weight_status,
            calc_continue,
        };

        if (chunk_type_id === 32) {
            // NJD_CV_SH
            cursor.seek(4); // Always 1.0
        } else if (chunk_type_id === 33) {
            // NJD_CV_VN_SH
            cursor.seek(4); // Always 1.0
            vertex.normal = new Vec3(
                cursor.f32(), // x
                cursor.f32(), // y
                cursor.f32() // z
            );
            cursor.seek(4); // Always 0.0
        } else if (35 <= chunk_type_id && chunk_type_id <= 40) {
            if (chunk_type_id === 37) {
                // NJD_CV_NF
                // NinjaFlags32
                vertex.index = index + cursor.u16();
                vertex.bone_weight = cursor.u16() / 255;
            } else {
                // Skip user flags and material information.
                cursor.seek(4);
            }
        } else if (41 <= chunk_type_id && chunk_type_id <= 47) {
            vertex.normal = new Vec3(
                cursor.f32(), // x
                cursor.f32(), // y
                cursor.f32() // z
            );

            if (chunk_type_id >= 42) {
                if (chunk_type_id === 44) {
                    // NJD_CV_VN_NF
                    // NinjaFlags32
                    vertex.index = index + cursor.u16();
                    vertex.bone_weight = cursor.u16() / 255;
                } else {
                    // Skip user flags and material information.
                    cursor.seek(4);
                }
            }
        } else if (48 <= chunk_type_id && chunk_type_id <= 50) {
            // 32-Bit vertex normal in format: reserved(2)|x(10)|y(10)|z(10)
            const normal = cursor.u32();
            vertex.normal = new Vec3(
                ((normal >> 20) & 0x3ff) / 0x3ff,
                ((normal >> 10) & 0x3ff) / 0x3ff,
                (normal & 0x3ff) / 0x3ff
            );

            if (chunk_type_id >= 49) {
                // Skip user flags and material information.
                cursor.seek(4);
            }
        }

        vertices.push(vertex);
    }

    return vertices;
}

function parse_triangle_strip_chunk(
    cursor: BufferCursor,
    chunk_type_id: number,
    flags: number
): NjTriangleStrip[] {
    const render_flags = {
        ignore_light: (flags & 0b1) !== 0,
        ignore_specular: (flags & 0b10) !== 0,
        ignore_ambient: (flags & 0b100) !== 0,
        use_alpha: (flags & 0b1000) !== 0,
        double_side: (flags & 0b10000) !== 0,
        flat_shading: (flags & 0b100000) !== 0,
        environment_mapping: (flags & 0b1000000) !== 0,
    };
    const user_offset_and_strip_count = cursor.u16();
    const user_flags_size = user_offset_and_strip_count >>> 14;
    const strip_count = user_offset_and_strip_count & 0x3fff;
    let options;

    switch (chunk_type_id) {
        case 64:
            options = [false, false, false, false];
            break;
        case 65:
            options = [true, false, false, false];
            break;
        case 66:
            options = [true, false, false, false];
            break;
        case 67:
            options = [false, false, true, false];
            break;
        case 68:
            options = [true, false, true, false];
            break;
        case 69:
            options = [true, false, true, false];
            break;
        case 70:
            options = [false, true, false, false];
            break;
        case 71:
            options = [true, true, false, false];
            break;
        case 72:
            options = [true, true, false, false];
            break;
        case 73:
            options = [false, false, false, false];
            break;
        case 74:
            options = [true, false, false, true];
            break;
        case 75:
            options = [true, false, false, true];
            break;
        default:
            throw new Error(`Unexpected chunk type ID: ${chunk_type_id}.`);
    }

    const [parse_texture_coords, parse_color, parse_normal, parse_texture_coords_hires] = options;

    const strips: NjTriangleStrip[] = [];

    for (let i = 0; i < strip_count; ++i) {
        const winding_flag_and_index_count = cursor.i16();
        const clockwise_winding = winding_flag_and_index_count < 1;
        const index_count = Math.abs(winding_flag_and_index_count);

        const vertices: NjMeshVertex[] = [];

        for (let j = 0; j < index_count; ++j) {
            const vertex: NjMeshVertex = {
                index: cursor.u16(),
            };
            vertices.push(vertex);

            if (parse_texture_coords) {
                cursor.seek(4);
            }

            if (parse_color) {
                cursor.seek(4);
            }

            if (parse_normal) {
                vertex.normal = new Vec3(cursor.u16(), cursor.u16(), cursor.u16());
            }

            if (parse_texture_coords_hires) {
                cursor.seek(8);
            }

            if (j >= 2) {
                cursor.seek(2 * user_flags_size);
            }
        }

        strips.push({
            ...render_flags,
            clockwise_winding,
            vertices,
        });
    }

    return strips;
}

File: lib/src/commonMain/kotlin/world/phantasmal/lib/fileFormats/ninja/Nj.kt
package world.phantasmal.lib.fileFormats.ninja

import mu.KotlinLogging
import world.phantasmal.core.isBitSet
import world.phantasmal.lib.cursor.Cursor
import world.phantasmal.lib.fileFormats.Vec2
import world.phantasmal.lib.fileFormats.Vec3
import world.phantasmal.lib.fileFormats.vec3Float
import kotlin.math.abs

// TODO:
//  - colors
//  - bump maps

private val logger = KotlinLogging.logger {}

// TODO: Simplify parser by not parsing chunks into vertices and meshes. Do the chunk to vertex/mesh
//       conversion at a higher level.
fun parseNjModel(cursor: Cursor, cachedChunks: MutableMap<Int, List<NjChunk>>): NjModel {
    val vlistOffset = cursor.int() // Vertex list
    val plistOffset = cursor.int() // Triangle strip index list
    val collisionSphereCenter = cursor.vec3Float()
    val collisionSphereRadius = cursor.float()
    val vertices: MutableList<NjVertex?> = mutableListOf()
    val meshes: MutableList<NjTriangleStrip> = mutableListOf()

    if (vlistOffset != 0) {
        cursor.seekStart(vlistOffset)

        for (chunk in parseChunks(cursor)) {
            if (chunk is NjChunk.Vertex) {
                for (vertex in chunk.vertices) {
                    while (vertices.size <= vertex.index) {
                        vertices.add(null)
                    }

                    vertices[vertex.index] = NjVertex(
                        vertex.position,
                        vertex.normal,
                        vertex.boneWeight,
                        vertex.boneWeightStatus,
                        vertex.calcContinue,
                    )
                }
            }
        }
    }

    if (plistOffset > 0) {
        cursor.seekStart(plistOffset)

        PolygonChunkProcessor(cachedChunks, meshes).process(parseChunks(cursor))
    }

    return NjModel(
        vertices,
        meshes,
        collisionSphereCenter,
        collisionSphereRadius,
    )
}

private class PolygonChunkProcessor(
    private val cachedChunks: MutableMap<Int, List<NjChunk>>,
    private val meshes: MutableList<NjTriangleStrip>,
) {
    private var textureId: Int? = null
    private var srcAlpha: Int? = null
    private var dstAlpha: Int? = null

    /**
     * When [cacheList] is non-null we are caching chunks.
     */
    private var cacheList: MutableList<NjChunk>? = null

    fun process(chunks: List<NjChunk>) {
        for (chunk in chunks) {
            if (cacheList == null) {
                when (chunk) {
                    is NjChunk.BlendAlpha -> {
                        srcAlpha = chunk.srcAlpha
                        dstAlpha = chunk.dstAlpha
                    }

                    is NjChunk.CachePolygonList -> {
                        cacheList = mutableListOf()
                        cachedChunks[chunk.cacheIndex] = cacheList!!
                    }

                    is NjChunk.DrawPolygonList -> {
                        val cached = cachedChunks[chunk.cacheIndex]

                        if (cached == null) {
                            logger.debug {
                                "Draw Polygon List chunk pointed to nonexistent cache index ${chunk.cacheIndex}."
                            }
                        } else {
                            process(cached)
                        }
                    }

                    is NjChunk.Tiny -> {
                        textureId = chunk.textureId
                    }

                    is NjChunk.Material -> {
                        srcAlpha = chunk.srcAlpha
                        dstAlpha = chunk.dstAlpha
                    }

                    is NjChunk.Strip -> {
                        for (strip in chunk.triangleStrips) {
                            strip.textureId = textureId
                            strip.srcAlpha = srcAlpha
                            strip.dstAlpha = dstAlpha
                        }

                        meshes.addAll(chunk.triangleStrips)
                    }

                    else -> {
                        // Ignore
                    }
                }
            } else {
                cacheList!!.add(chunk)
            }
        }
    }
}

private fun parseChunks(cursor: Cursor): List<NjChunk> {
    val chunks: MutableList<NjChunk> = mutableListOf()

    do {
        val chunkStartPosition = cursor.position
        val typeId = cursor.uByte().toInt()
        val flags = cursor.uByte().toInt()
        val chunkDataPosition = cursor.position
        var size = 0
        val chunk: NjChunk

        when (typeId) {
            0 -> {
                chunk = NjChunk.Null
            }
            1 -> {
                chunk = NjChunk.BlendAlpha(
                    srcAlpha = (flags ushr 3) and 0b111,
                    dstAlpha = flags and 0b111,
                )
            }
            2 -> {
                chunk = NjChunk.MipmapDAdjust(
                    adjust = flags and 0b1111,
                )
            }
            3 -> {
                chunk = NjChunk.SpecularExponent(
                    specular = flags and 0b11111,
                )
            }
            4 -> {
                chunk = NjChunk.CachePolygonList(
                    cacheIndex = flags,
                )
            }
            5 -> {
                chunk = NjChunk.DrawPolygonList(
                    cacheIndex = flags,
                )
            }
            in 8..9 -> {
                size = 2
                val textureBitsAndId = cursor.uShort().toInt()

                chunk = NjChunk.Tiny(
                    typeId,
                    flipU = flags.isBitSet(7),
                    flipV = flags.isBitSet(6),
                    clampU = flags.isBitSet(5),
                    clampV = flags.isBitSet(4),
                    mipmapDAdjust = flags.toUInt() and 0b1111u,
                    filterMode = textureBitsAndId ushr 14,
                    superSample = (textureBitsAndId and 0x40) != 0,
                    textureId = textureBitsAndId and 0x1FFF,
                )
            }
            in 17..31 -> {
                val bodySize = 2 * cursor.short()
                size = 2 + bodySize

                var diffuse: NjArgb? = null
                var ambient: NjArgb? = null
                var specular: NjErgb? = null

                if (typeId == 24) {
                    // Skip bump map data.
                    cursor.seek(bodySize)
                } else {
                    if (typeId.isBitSet(0)) {
                        diffuse = NjArgb(
                            b = cursor.uByte().toFloat() / 255f,
                            g = cursor.uByte().toFloat() / 255f,
                            r = cursor.uByte().toFloat() / 255f,
                            a = cursor.uByte().toFloat() / 255f,
                        )
                    }

                    if (typeId.isBitSet(1)) {
                        ambient = NjArgb(
                            b = cursor.uByte().toFloat() / 255f,
                            g = cursor.uByte().toFloat() / 255f,
                            r = cursor.uByte().toFloat() / 255f,
                            a = cursor.uByte().toFloat() / 255f,
                        )
                    }

                    if (typeId.isBitSet(2)) {
                        specular = NjErgb(
                            b = cursor.uByte(),
                            g = cursor.uByte(),
                            r = cursor.uByte(),
                            e = cursor.uByte(),
                        )
                    }
                }

                chunk = NjChunk.Material(
                    typeId,
                    srcAlpha = (flags ushr 3) and 0b111,
                    dstAlpha = flags and 0b111,
                    diffuse,
                    ambient,
                    specular,
                )
            }
            in 32..50 -> {
                size = 2 + 4 * cursor.short()
                chunk = NjChunk.Vertex(
                    typeId,
                    vertices = parseVertexChunk(cursor, typeId, flags),
                )
            }
            in 56..58 -> {
                size = 2 + 2 * cursor.short()
                chunk = NjChunk.Volume(
                    typeId,
                )

                // Skip volume information.
                cursor.seek(2 * cursor.short())
            }
            in 64..75 -> {
                size = 2 + 2 * cursor.short()
                chunk = NjChunk.Strip(
                    typeId,
                    triangleStrips = parseTriangleStripChunk(cursor, typeId, flags),
                )
            }
            255 -> {
                chunk = NjChunk.End
            }
            else -> {
                val bodySize = 2 * cursor.short()
                size = 2 + bodySize
                chunk = NjChunk.Unknown(
                    typeId,
                )
                // Skip unknown data.
                cursor.seek(bodySize)
                logger.warn { "Unknown chunk type $typeId at offset ${chunkStartPosition}." }
            }
        }

        chunks.add(chunk)

        val bytesRead = cursor.position - chunkDataPosition

        check(bytesRead <= size) {
            "Expected to read at most $size bytes, actually read $bytesRead."
        }

        cursor.seekStart(chunkDataPosition + size)
    } while (chunk != NjChunk.End)

    return chunks
}

private fun parseVertexChunk(
    cursor: Cursor,
    chunkTypeId: Int,
    flags: Int,
): List<NjChunkVertex> {
    val boneWeightStatus = flags and 0b11
    val calcContinue = (flags and 0x80) != 0

    val index = cursor.uShort()
    val vertexCount = cursor.uShort()

    val vertices: MutableList<NjChunkVertex> = mutableListOf()

    for (i in (0u).toUShort() until vertexCount) {
        var vertexIndex = index + i
        val position = cursor.vec3Float()
        var normal: Vec3? = null
        var boneWeight: Float? = null

        when (chunkTypeId) {
            32 -> {
                // NJD_CV_SH
                cursor.seek(4) // Always 1.0
            }
            33 -> {
                // NJD_CV_VN_SH
                cursor.seek(4) // Always 1.0
                normal = cursor.vec3Float()
                cursor.seek(4) // Always 0.0
            }
            34 -> {
                // NJD_CV
                // Nothing to do.
            }
            in 35..40 -> {
                if (chunkTypeId == 37) {
                    // NJD_CV_NF
                    // NinjaFlags32
                    vertexIndex = index + cursor.uShort()
                    boneWeight = cursor.uShort().toFloat() / 255f
                } else {
                    // NJD_CV_D8
                    // NJD_CV_UF
                    // NJD_CV_S5
                    // NJD_CV_S4
                    // NJD_CV_IN
                    // Skip user flags and material information.
                    cursor.seek(4)
                }
            }
            41 -> {
                // NJD_CV_VN
                normal = cursor.vec3Float()
            }
            in 42..47 -> {
                normal = cursor.vec3Float()

                if (chunkTypeId == 44) {
                    // NJD_CV_VN_NF
                    // NinjaFlags32
                    vertexIndex = index + cursor.uShort()
                    boneWeight = cursor.uShort().toFloat() / 255f
                } else {
                    // NJD_CV_VN_D8
                    // NJD_CV_VN_UF
                    // NJD_CV_VN_S5
                    // NJD_CV_VN_S4
                    // NJD_CV_VN_IN
                    // Skip user flags and material information.
                    cursor.seek(4)
                }
            }
            in 48..50 -> {
                // NJD_CV_VNX
                // 32-Bit vertex normal in format: reserved(2)|x(10)|y(10)|z(10)
                val n = cursor.uInt()
                normal = Vec3(
                    ((n shr 20) and 0x3FFu).toFloat() / 0x3FF,
                    ((n shr 10) and 0x3FFu).toFloat() / 0x3FF,
                    (n and 0x3FFu).toFloat() / 0x3FF,
                )

                if (chunkTypeId >= 49) {
                    // NJD_CV_VNX_D8
                    // NJD_CV_VNX_UF
                    // Skip user flags and material information.
                    cursor.seek(4)
                }
            }
            else -> error("Unexpected chunk type ID ${chunkTypeId}.")
        }

        vertices.add(NjChunkVertex(
            vertexIndex.toInt(),
            position,
            normal,
            boneWeight,
            boneWeightStatus,
            calcContinue,
        ))
    }

    return vertices
}

private fun parseTriangleStripChunk(
    cursor: Cursor,
    chunkTypeId: Int,
    flags: Int,
): List<NjTriangleStrip> {
    val ignoreLight = flags.isBitSet(0)
    val ignoreSpecular = flags.isBitSet(1)
    val ignoreAmbient = flags.isBitSet(2)
    val useAlpha = flags.isBitSet(3)
    val doubleSide = flags.isBitSet(4)
    val flatShading = flags.isBitSet(5)
    val environmentMapping = flags.isBitSet(6)

    val userOffsetAndStripCount = cursor.short().toInt()
    val userFlagsSize = 2 * (userOffsetAndStripCount ushr 14)
    val stripCount = userOffsetAndStripCount and 0x3FFF

    var hasTexCoords = false
    var hasColor = false
    var hasNormal = false
    var hasDoubleTexCoords = false

    when (chunkTypeId) {
        64 -> {
        }
        65, 66 -> {
            hasTexCoords = true
        }
        67 -> {
            hasNormal = true
        }
        68, 69 -> {
            hasTexCoords = true
            hasNormal = true
        }
        70 -> {
            hasColor = true
        }
        71, 72 -> {
            hasTexCoords = true
            hasColor = true
        }
        73 -> {
        }
        74, 75 -> {
            hasDoubleTexCoords = true
        }
        else -> error("Unexpected chunk type ID: ${chunkTypeId}.")
    }

    val strips: MutableList<NjTriangleStrip> = mutableListOf()

    repeat(stripCount) {
        val windingFlagAndIndexCount = cursor.short().toInt()
        val clockwiseWinding = windingFlagAndIndexCount < 0
        val indexCount = abs(windingFlagAndIndexCount)

        val vertices: MutableList<NjMeshVertex> = mutableListOf()

        for (j in 0 until indexCount) {
            val index = cursor.uShort().toInt()

            val texCoords = if (hasTexCoords) {
                Vec2(cursor.uShort().toFloat() / 255f, cursor.uShort().toFloat() / 255f)
            } else null

            // Ignore ARGB8888 color.
            if (hasColor) {
                cursor.seek(4)
            }

            val normal = if (hasNormal) {
                Vec3(
                    cursor.uShort().toFloat() / 255f,
                    cursor.uShort().toFloat() / 255f,
                    cursor.uShort().toFloat() / 255f,
                )
            } else null

            // Ignore double texture coordinates (Ua, Vb, Ua, Vb).
            if (hasDoubleTexCoords) {
                cursor.seek(8)
            }

            // User flags start at the third vertex because they are per-triangle.
            if (j >= 2) {
                cursor.seek(userFlagsSize)
            }

            vertices.add(NjMeshVertex(
                index,
                normal,
                texCoords,
            ))
        }

        strips.add(NjTriangleStrip(
            ignoreLight,
            ignoreSpecular,
            ignoreAmbient,
            useAlpha,
            doubleSide,
            flatShading,
            environmentMapping,
            clockwiseWinding,
            hasTexCoords,
            hasNormal,
            textureId = null,
            srcAlpha = null,
            dstAlpha = null,
            vertices,
        ))
    }

    return strips
}
